plugins/processors/reversedns/rdnscache.go

```diff
@@ -0,0 +1,324 @@
+	return d.lookup(ip)
+}
+
+func (d *ReverseDNSCache) lookup(ip string) []string {
+	// check if the value is cached
+	d.rwLock.RLock()
+	result, found := d.unsafeGetFromCache(ip)
+	if found && result.completed && result.expiresAt.After(time.Now()) {
+		debug("found in cache", result)
+		defer d.rwLock.RUnlock()
+		atomic.AddUint64(&d.stats.CacheHit, 1)
+		// cache is valid
+		debug("Found in cache: ", result.domains)
+		return result.domains
+	}
+	d.rwLock.RUnlock()
+
+	// if it's not cached, kick off a lookup job and subscribe to the result.
+	debug("subscribe!")
+	lookupChan := d.subscribeTo(ip)
+	timer := time.NewTimer(d.lookupTimeout)
+	defer timer.Stop()
+
+	// timer is still necessary even if doLookup respects timeout due to worker
+	// pool starvation.
+	select {
+	case domains := <-lookupChan:
+		return domains
+	case <-timer.C:
+		fmt.Fprintf(os.Stderr, "reverse dns lookup timed out\n")
+		return []string{}
+	}
+}
+
+func (d *ReverseDNSCache) subscribeTo(ip string) callbackChannelType {
+	callback := make(callbackChannelType, 1)
+
+	d.rwLock.Lock()
+	defer d.rwLock.Unlock()
+	debug("d.rwLock.Lock()")
+	defer debug("d.rwLock.Unlock()")
+
+	// confirm it's still not in the cache. This needs to be done under an active lock.
+	result, found := d.unsafeGetFromCache(ip)
+	if found {
+		debug("found")
+		atomic.AddUint64(&d.stats.CacheHit, 1)
+		// has the request been answered since we last checked?
+		if result.completed {
+			debug("found completed")
+			// we can return the answer with the channel.
+			callback <- result.domains
+			return callback
+		}
+		debug("found not completed")
+		// there's a request but it hasn't been answered yet;
+		// add yourself to the subscribers and return that.
+		result.callbacks = append(result.callbacks, callback)
+		d.unsafeSaveToCache(result)
+		return callback
+	}
+
+	atomic.AddUint64(&d.stats.CacheMiss, 1)
+
+	debug("not found")
+
+	// otherwise we need to register the request
+	l := &dnslookup{
+		ip:        ip,
+		expiresAt: time.Now().Add(d.ttl),
+		callbacks: []callbackChannelType{callback},
+	}
+
+	d.unsafeSaveToCache(l)
+	go d.doLookup(*l)
+	return callback
+}
+
+// unsafeGetFromCache fetches from the correct internal ip cache.
+// you MUST first do a read or write lock before calling it, and keep locks around
+// the dnslookup that is returned until you clone it.
+func (d *ReverseDNSCache) unsafeGetFromCache(ip string) (lookup *dnslookup, found bool) {
+	lookup, found = d.cache[ip]
+	if found && lookup.expiresAt.Before(time.Now()) {
+		return nil, false
+	}
+	return lookup, found
+}
+
+// unsafeSaveToCache stores a lookup in the correct internal ip cache.
+// you MUST first do a read or write lock before calling it.
+func (d *ReverseDNSCache) unsafeSaveToCache(lookup *dnslookup) {
+	if lookup.expiresAt.Before(time.Now()) {
+		return // don't cache.
+	}
+	d.cache[lookup.ip] = lookup
+}
+
+func (d *ReverseDNSCache) startCleanupWorker(ctx context.Context) {
+	go func() {
```

Comment:
I think I considered this but was concerned that it could introduce new problems, like overlapping cleanup runs. I like getting rid of the Stop though... might end up with a channel instead to signal once every ... 10k requests?